src/control/controlvalue.h

```diff
@@ -79,7 +102,24 @@ class ControlValueAtomicBase {
             // m_currentIndex and in the mean while a reader locks the formally current value
             // because it has written cRingSize times. Reading the less recent value will fix
             // it because it is now actually the current value.
-            index = (index - 1) % (cRingSize);
+            index = (index - 1) % cRingSize;
+        }
+        return value;
+    }
+
+    inline T getValueOnce() {
+        T value;
+        unsigned int index = static_cast<unsigned int>(load_atomic(m_readIndex)) % cRingSize;
+        while (!m_ring[index].tryGetOnce(&value)) {
+            // We are here if
```

Comment:
ControlValue is conceptually used as a FIFO queue with 1 writer, 1 reader, and a size of 1. Just a single slot with move semantics.

Currently multiple readers would indeed be able to read the same value twice by accessing the same slot before moving the value out. What we actually need here is the concept of channels like we have in Rust.

We cannot use std::atomic<std::shared_ptr<T>>, because this is not guaranteed to be atomic. I'm also unsure about which version of the C++ standard actually includes the partial template specialization?

Reply:
QAtomicPointer does not allow to use a custom deleter. But maybe  we can wrap it and ensure that any stored value gets "deleted" upon destruction.

Reply:
The straight forward solution of transforming `submitNextTrack()` into a slot that is triggered by `QMetaObject::invokeMethod()` does not work here, because the design of `WorkerThread` forbids using an event loop.

I agree that the `...Once` functions in `ControlValueAtomic` methods only work as expected for this special use case of both a single reader and writer. In https://github.com/uklotzde/mixxx/tree/2.2_aoide I'm using a pure Qt approach for worker threads that works very well. But currently I'm not motivated to invest a lot of time into transforming the whole design.

Reply:
Ok, I found a loophole and reverted the changes for ControlValue. Replaced it with a new, lock-free data structure that is more appropriate here.